### PR TITLE
(aws_credentials): update ecr url validation regex

### PR DESCRIPTION
**Pull request by dims**
Updates the regex for ECR URL validation to support isolated regions and includes additional testcases for these.

Trying to reduce patches carried in eks-distro:
- https://github.com/aws/eks-distro/blob/main/projects/kubernetes/kubernetes/1-24/patches/0004-EKS-PATCH-aws_credentials-update-ecr-url-validation-.patch

Signed-off-by: Jyoti Mahapatra <jyotima@amazon.com>
Signed-off-by: Davanum Srinivas <davanum@gmail.com>

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ import (
 )
 
 var (
-	ecrPattern = regexp.MustCompile(`^(\d{12})\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?$`)
+	ecrPattern = regexp.MustCompile(`^(\d{12})\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.(amazonaws\.com(\.cn)?|sc2s\.sgov\.gov|c2s\.ic\.gov)$`)
```
Comment (liggitt) [inline pkg/credentialprovider/aws/aws_credentials.go:44]: @andrewsykim can you confirm the credential provider plugin can handle logic like this?

Comment (dims) [reply pkg/credentialprovider/aws/aws_credentials.go:44]: @liggitt we've had this patch in prod for quite a bit of time from the looks of it.

- https://github.com/aws/aws-sdk-js/blame/master/lib/region_config_data.json#L27-L36
- https://github.com/hashicorp/terraform-provider-aws/issues/19000
- https://github.com/hashicorp/terraform-provider-aws/issues/19014
- https://github.com/hashicorp/terraform-provider-aws/issues/9993

Funnily we have it in our documentation too :)  https://kubernetes.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/

Comment (liggitt) [reply pkg/credentialprovider/aws/aws_credentials.go:44]: > Funnily we have it in our documentation too :) https://kubernetes.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/

ha (though that example is more constrained than the entire subdomain .. only ".dkr.ecr.us-iso-east-1.c2s.ic.gov" and ".dkr.ecr.us-isob-east-1.sc2s.sgov.gov" suffixes, which are much more clearly ecr

Comment (andrewsykim) [reply pkg/credentialprovider/aws/aws_credentials.go:44]: > Funnily we have it in our documentation too :) https://kubernetes.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/

This is documentation for the external credential provider though, which is to demonstrate that you can configure match images with whatever images you want and removes the need for us to continuously update the internal plugin with new subdomains.

**Pull request merged**